lib/vfscore: Implement individual volume automounting

mogasergiu opened the pull request:
Provide a command-line argument `vfs.fstab` that is meant to contain a list of whitespace separated strings with the following format:
```
	vfs.fstab=[
		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
		...
	]
```

The core will parse the provided strings as volume information and attempt to mount them accordingly.

<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

mogasergiu committed:
lib/vfscore: Implement individual volume automounting

Provide a command-line argument `vfs.fstab` that is meant to
contain a list of whitespace separated strings with the following
format:
```
	vfs.fstab=[
		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
		...
	]
```

The core will parse the provided strings as volume information and
attempt to mount them accordingly, if `CONFIG_LIBVFSCORE_FSTAB` is
provided.

Furthermore, add a configurable `CONFIG_LIBVFSCORE_FSTAB_SIZE`
option that will be used as the maximum amount of automatically mounted
volumes, excluding the `rootfs` and modularize `initrd` extraction and
mounting functionality into a standalone function to be used by both
`fstab` and implicit `rootfs`.

As a last change, since the file's functionality drastically changed,
rename `rootfs.c` to `automount.c`.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

skuenzer commented on lib/vfscore/Makefile.uk:
```diff
@@ -10,6 +10,7 @@ LIBVFSCORE_SRCS-y += $(LIBVFSCORE_BASE)/file.c
 LIBVFSCORE_SRCS-y += $(LIBVFSCORE_BASE)/stdio.c
 LIBVFSCORE_SRCS-y += $(LIBVFSCORE_BASE)/mount.c
 LIBVFSCORE_SRCS-y += $(LIBVFSCORE_BASE)/vnode.c
+LIBVFSCORE_SRCS-y += $(LIBVFSCORE_BASE)/volumes.c
```
Can we name the file `automount.c`? I think as a little clean-up it makes sense to rename `rootfs.c` and then add your code on top. Then we have it within one file.

mogasergiu replied:
Are you sure? Because that would make volumes depend on `CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS`. And there is a case where you may want to have atuomatic mount of volumes without having the `rootfs.c` functionality.

skuenzer replied:
Both true but you could just guard the ROOTFS code in the file and keep the source file part of the build, independent of the switch, right? My point is that it is easier to lookup any automount related code and that it is close together from an organization perspective.

mogasergiu replied:
So guard two thirds of the file? Does it not look like bad design since the separation is so big to not have them in separate files so that the separation is transparent from the source code point of view?

skuenzer commented on lib/vfscore/volumes.c:
```diff
@@ -0,0 +1,59 @@
+#include <uk/init.h>
+#include <uk/libparam.h>
+
+#define UK_MAX_VFSCORE_VOLUMES_COUNT			64
```
Can we make this available via menu config? Also make it configurable if automount is available?

skuenzer commented on lib/vfscore/volumes.c:
```diff
@@ -0,0 +1,59 @@
+}
+
+/* Should be preferrably called after vfscore's rootfs is mounted (prio 4) */
+uk_rootfs_initcall_prio(vfscore_volumes_mount, 5);
```
great! It is a bit clearer maybe if we have the registrations in the same file. We could also include your code to the rootfs mount code.

StefanJum reviewed (commented):
This seems to work fine @mogasergiu, I'll add the tag after you resolve Simon's comments.

RaduNichita commented on lib/vfscore/volumes.c:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Expected command-line argument format:
+ *	vfs.automount=[
+ *		"<source_device>:<mount_target>:<fs_driver>:<options>"
```
```suggestion
 *		"<source_device>:<mount_target>:<fs_driver>:<options>",
```
I think there should be a comma here.

mogasergiu replied:
@skuenzer AFAIK, the `uklibparam` design makes use of whitespaces in such cases :thinking:.

skuenzer replied:
Yes, no commas. This has actually a pragmatic reason because that missing comma saves us parsing code. We can just rely on `ukargparse` for doing the argument split and array parsing. `ukargparse` also handles quotes and escaped quotes correctly. `uklibparam` automatically benefits from this.

RaduNichita commented on lib/vfscore/volumes.c:
```diff
@@ -0,0 +1,59 @@
+	return 0;
+}
+
+/* Should be preferrably called after vfscore's rootfs is mounted (prio 4) */
```
Could you elaborate a little bit about priorities. Why it should be 4 here?

mogasergiu replied:
What I meant was that `rootfs.c` has a `uk_rootfs_initcall_prio(vfscore_rootfs, 4);`. Therefore, in order to ensure this can only be called after an explicit rootfs, give this a priority of `4 + 1 = 5`. I was referring to the `rootfs`. I will make the comment more clear :100:.

skuenzer replied:
I think if the `uk_rootfs_initcall_prio()` calls are next to each other (I am speaking about merged `rootfs.c` and `volumes.c`), then this is also clear. Additionally, `<uk/prio.h>` provides some computation macros so that this could be by using those also be a bit more expressive by the code. And the last alternative I see is to use only a single automount function that handles first `rootfs` and then `volumes`...

RaduNichita reviewed (changes requested):
Great work, @mogasergiu :muscle: . I left some minor nitpicks.
I will approve after that!

StefanJum reviewed (commented):
I retested this and everything works fine, thanks!
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

skuenzer commented on lib/vfscore/automount.c:
```diff
@@ -123,8 +155,33 @@ static int vfscore_rootfs(void)
+	return 0;
+}
+
+static int vfscore_volumes_automount(void)
```
This is more a minor detail, more aesthetically: I'd prefer two functions at the same level:
`vfscore_automount_rootfs`
`vfscore_automount_volumes`
and a third function that calls these two and is registered to the init table.
btw, it is a clever idea to create empty function bodies depending on the configuration.

RaduNichita reviewed (approved):
All good now. Thanks, @mogasergiu :1st_place_medal: 

Reviewed-by: Radu Nichita <radunichita99@gmail.com>

skuenzer commented on lib/vfscore/Config.uk:
```diff
@@ -100,5 +100,19 @@ if LIBVFSCORE_AUTOMOUNT_ROOTFS
 		filesystem.
 endif
 
+config LIBVFSCORE_MAX_VOLUMES_COUNT
```
Can we also introduce an option that switches this on or off? I am thinking of something along these lines:
```
config LIBVFSCORE_AUTOMOUNT_VOLUMES
bool "Automatically mount list of volumes from kernel arguments"
selects LIBUKLIBPARAM
default n

if LIBVFSCORE_AUTOMOUNT_VOLUMES
    config LIBVFSCORE_AUTOMOUNT_VOLUMES_COUNT
    # ...
endif
```

StefanJum reviewed (commented):
Sill no bugs found in the wild :sunglasses: 
The only thing I would do is move the `help` section from under the `LIBVFSCORE_MAX_VOLUMES_COUNT` to the `LIBVFSCORE_AUTOMOUNT_VOLUMES` option. It's not really related to the max number of volumes.

razvanvirtan reviewed (approved):
All good for me, thanks!
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

mogasergiu commented:
> Sill no bugs found in the wild sunglasses The only thing I would do is move the `help` section from under the `LIBVFSCORE_MAX_VOLUMES_COUNT` to the `LIBVFSCORE_AUTOMOUNT_VOLUMES` option. It's not really related to the max number of volumes.

Good point, I simply copy-pasted. :rofl:

StefanJum reviewed (approved):
All good now
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

mogasergiu committed:
lib/vfscore: Discontinue `vfs.{rootfs, rootdef, rootops, rootflags}`

Now that we can dynamically mount any kind of volume through a more
generic format through the commandline, there is an unnecessary
overlap between `vfs.{rootfs, rootdef, rootops,rootflags}` and
defining `rootfs` related arguments through `vfs.fstab` volumes.

Therefore, deprecate `vfs.{rootfs, rootdef, rootops,rootflags}` in
favor of `vfs.fstab`.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu commented on lib/vfscore/automount.c:
```diff
@@ -0,0 +1,309 @@
+#else /* CONFIG_LIBUKCPIO */
+static int vfscore_mount_initrd_volume(struct vfscore_volume *vv __unused)
+{
+	uk_pr_warn("Cannot extract InitRD. Enable CONFIG_LIBRAMFS.\n");
```
Oops should have been `CONFIG_LIBUKCPIO`

skuenzer replied:
Probably we should just fall back to a default message, something like saying: `Unknown filesystem: initrd`. I guess we have something like this already built-in with `mount()`. So, I think a separate handler for "initrd-support-is-not-present" is not needed.

mogasergiu replied:
Right then, I will just let this be an empty stub.

skuenzer commented on lib/vfscore/automount.c:
```diff
@@ -0,0 +1,309 @@
+	for (i = 0; i < CONFIG_LIBVFSCORE_FSTAB_SIZE && vfscore_fstab[i]; i++) {
+		vfscore_fstab_fetch_volume_args(vfscore_fstab[i], &vv);
+
+		uk_pr_info("Mount %s to %s...\n", vv.sdev, vv.path);
```
What is `vv.sdev` saying in the initrd case? Maybe you could place this message under the else case and print a message for initrd like: "Extracting initrd to %s...\n"

mogasergiu replied:
`vv.sdev` is a mandatory argument as per `fstab` syntax:

```
vfs.fstab=[
		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
		"<src_dev>:<mntpoint>:<fsdriver>[:<flags>:<opts>]"
		...
	]
```
`vv.sdev` is therefore `<src_dev>`.

EDIT: The initrd specific printing is already done by `vfscore_mount_initrd_volume`

skuenzer replied:
My question was that the initrd mounting code does not need a source device as parameter. According to the code we accept anything that is in there which is actually fine. :+1:

skuenzer commented on lib/vfscore/automount.c:
```diff
@@ -0,0 +1,312 @@
+		return -1;
+	}
+
+	uk_pr_info("Mount ramfs to /...\n");
```
I think this message is not needed anymore and it is incorrect. initrd can go now to any submountpoint, right?

mogasergiu replied:
Now that you mention it, [mount](https://github.com/unikraft/unikraft/blob/staging/lib/vfscore/mount.c#L122) already has a similar message. Therefore the message should be dropped from `vfscore_automount_fstab_volumes` as well :+1:.

skuenzer commented on lib/vfscore/automount.c:
```diff
@@ -0,0 +1,312 @@
+	rc = ukplat_memregion_find_initrd0(&initrd);
+	if (unlikely(rc < 0)) {
+		uk_pr_crit("Could not find an initrd!\n");
+
```
I think these spaces between message and return code is not necessary for these small if blocks. I leave it up to you if you want to keep them or not.

skuenzer commented:
I just noticed, that in your second commit that removes the `vfs.roofs` parameter, etc. you also need to update `Config.uk`:
```diff
 config LIBVFSCORE_AUTOMOUNT_ROOTFS
 bool "Automatically mount a root filesysytem (/)"
 default n
-imply LIBUKLIBPARAM
 help
-	Automatically mounts '/' during boot. If `libuklibparam` is
-	compiled in, the default root filesystem and mount options can
-	be changed with the following library parameters:
-	'vfs.rootfs', 'vfs.rootdev', 'vfs.rootflags', and 'vfs.rootopts'
+	Automatically mounts '/' during boot. The parameters are
+	compiled-in and cannot be changed at runtime.
```

This is a minor comment: Instead of keeping the parameters set from config options from line 76 to 98, you could directly fill the struct in line 226 with those.

I aslo jsut notice that you are using `#ifdef CONFIG_ABC` although you can use `#if CONFIG_ABC` which is slightly more powerful because it catches the case when you manually do `#define CONFIG_ABC 0`. If you use `#if` on an undefined pre-processor macro, it will return false.

skuenzer commented on lib/vfscore/automount.c:
```diff
@@ -0,0 +1,307 @@
+			rc = vfscore_mount_volume(&vv);
+		if (unlikely(rc))
+			uk_pr_err("Failed to mount %s: error %d\n", vv.sdev,
+				  rc);
```
We should propagate failure as soon as one mount operation failed

mogasergiu commented:
> I aslo jsut notice that you are using `#ifdef CONFIG_ABC` although you can use `#if CONFIG_ABC` which is slightly more powerful because it catches the case when you manually do `#define CONFIG_ABC 0`. If you use `#if` on an undefined pre-processor macro, it will return false.

Is this just general advice? The problem is if you do `#if` on a string you are going to be met with a preprocessor error. So I think I will stick to `#ifdef`'s in the `CONFIG_LIBVFSCORE_ROOT*` macro's :thinking:.

skuenzer reviewed (approved):
Thanks a lot for the work!

Reviewed-by: Simon Kuenzer <simon@unikraft.io>
Approved-by: Simon Kuenzer <simon@unikraft.io>
